nltk/parse/dependencygraph.py

```diff
@@ -78,74 +97,64 @@ def add_arc(self, head_address, mod_address):
         node specified by the mod address.
         """
-        for node in self.nodelist:
-            if node['address'] == head_address and (mod_address not in node['deps']):
-                node['deps'].append(mod_address)
+        self.nodes[head_address]['deps'].append(mod_address)
 
     def connect_graph(self):
         """
         Fully connects all non-root nodes.  All nodes are set to be dependents
         of the root node.
         """
-        for node1 in self.nodelist:
-            for node2 in self.nodelist:
+        for node1 in self.nodes.values():
+            for node2 in self.nodes.values():
                 if node1['address'] != node2['address'] and node2['rel'] != 'TOP':
                     node1['deps'].append(node2['address'])
 
-    # fix error and return
     def get_by_address(self, node_address):
-        """
-        Returns the node with the given address.
-        """
-        for node in self.nodelist:
-            if node['address'] == node_address:
-                return node
-        print('THROW ERROR: address not found in -get_by_address-')
-        return -1
+        """Return the node with the given address."""
+        return self.nodes[node_address]
 
     def contains_address(self, node_address):
         """
         Returns true if the graph contains a node with the given node
         address, false otherwise.
         """
-        for node in self.nodelist:
-            if node['address'] == node_address:
-                return True
-        return False
+        return node_address in self.nodes
 
     def __str__(self):
-        return pformat(self.nodelist)
+        return pformat(self.nodes)
 
     def __repr__(self):
-        return "<DependencyGraph with %d nodes>" % len(self.nodelist)
+        return "<DependencyGraph with {0} nodes>".format(len(self.nodes))
 
     @staticmethod
-    def load(filename, zero_based=False):
+    def load(filename, zero_based=False, cell_separator=None):
         """
         :param filename: a name of a file in Malt-TAB format
-        :param zero_based: nodes in the input file are numbered starting from 0 rather
-            than 1 (as produced by, e.g., zpar)
+        :param zero_based: nodes in the input file are numbered starting from 0
+        rather than 1 (as produced by, e.g., zpar)
+        :param str cell_separator: the cell separator. If not provided, cells
+        are split by whitespace.
+
         :return: a list of DependencyGraphs
-        """
-        with open(filename) as infile:
-            return [DependencyGraph(tree_str, zero_based=zero_based) for tree_str in
-                                                  infile.read().split('\n\n')]
 
-    @staticmethod
-    def _normalize(line):
-        """
-        Deal with lines in which spaces are used rather than tabs.
         """
-        SPC = re.compile(' +')
-        return re.sub(SPC, '\t', line).strip()
+        with open(filename) as infile:
```

Comment:
I think file should be opened in text mode, we should get unicode data from it and be explicit about the required encoding. Now `tree_str` is a bytestring.